lib/elixir/lib/calendar/datetime.ex

```diff
@@ -206,31 +315,35 @@ defmodule DateTime do
       minute: minute,
       second: second,
       microsecond: microsecond,
-      std_offset: 0,
-      utc_offset: 0,
-      zone_abbr: "UTC",
-      time_zone: "Etc/UTC"
+      std_offset: std_offset,
+      utc_offset: utc_offset,
+      zone_abbr: zone_abbr,
+      time_zone: time_zone
     }
-
-    {:ok, datetime}
   end
 
   @doc """
   Converts the given `NaiveDateTime` to `DateTime`.
 
   It expects a time zone to put the NaiveDateTime in.
-  Currently it only supports "Etc/UTC" as time zone.
 
   ## Examples
 
       iex> DateTime.from_naive!(~N[2016-05-24 13:26:08.003], "Etc/UTC")
       #DateTime<2016-05-24 13:26:08.003Z>
 
+      iex> DateTime.from_naive!(~N[2018-05-24 13:26:08.003], "Europe/Copenhagen", FakeTimeZoneDatabase)
+      #DateTime<2018-05-24 13:26:08.003+02:00 CEST Europe/Copenhagen>
+
   """
   @doc since: "1.4.0"
-  @spec from_naive!(NaiveDateTime.t(), Calendar.time_zone()) :: t
-  def from_naive!(naive_datetime, time_zone) do
-    case from_naive(naive_datetime, time_zone) do
+  @spec from_naive!(
+          NaiveDateTime.t(),
+          Calendar.time_zone(),
+          TimeZoneDatabaseClient.tz_db_or_config()
+        ) :: t
+  def from_naive!(naive_datetime, time_zone, tz_db_or_config \\ :from_config) do
+    case from_naive(naive_datetime, time_zone, tz_db_or_config) do
```

Comment:
we also need to handle `{:ambiguous, _, _}` and `{:gap, _, _}` returns from `from_naive/3`, I guess we may have to raise on them.

Reply:
@lau what do you suggest here? Should we raise on ambiguous and gap?

Reply:
I think raising makes sense on both of those.

Here are some other ideas that are related:

`from_naive!` could raise on `:gap` because the provided naive datetime is not valid in the timezone. For `:ambiguous` though it could just return one of the valid results. In case people just want one of the valid results even though they won't know that there were two results.

I imagine that something that people might want in some cases is "just get me a valid datetime that is close enough if in a gap or ambiguous".  For instance if you get a `:gap` it just returns the second (or first) datetime of  `{:ambiguous, first_datetime, second_datetime}`. And the same thing for `{:gap, first_datetime, second_datetime}`. 

There could be a function where you pass the result of `from_naive/3` and it just returns either the only result if `{:ok, dt}`, or the second datetime of `{:gap, _, dt2}` or `{:ambiguous, _, dt2}`. e.g. `DateTime.from_naive(ndt, tz) |> DateTime.degap_and_disambiguate`. But maybe such a function is simple enough that it should not be in the standard library and could just be an example pattern matching that is in the documentation:
```
from_naive(...) ->
    {:ok, dt} -> {:ok, dt}
    {:gap, _, dt2} -> {:ok, dt2}
    {:ambiguous, _, dt2} -> {:ok, dt2}
    {:error, _} = error -> error
end
```

I'm not sure which of those I prefer or should be in the standard library, but some ideas to consider.

Reply:
Thanks @lau! Since I am not sure any of those choices are reasonable to pick as a default, let's make sure we raise for now and we can add new functions in the future.